### PR TITLE
Expose the list of completed ranges in SparseFileTracker

### PR DESCRIPTION
**Pull request by tlrx**
This PR adds a new method `getCompletedRanges()` to the `SparseFileTracker` class. This will be useful for the persistent cache that will need to retrieve the list of completed ranges before _fsyncing_ a cache file.

### CODE REVIEW

**Comment by original-brownbear** [comment]
@tlrx I'll take a look here shortly. Just FYI the test failure here is reproducible in master as well. I'm looking into it

**Comment by original-brownbear** [comment]
Test failure is fixed by https://github.com/elastic/elasticsearch/pull/64150

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java [hunk 1]
```diff
@@ -49,6 +50,23 @@ public long getLength() {
...
 
+    public List<Tuple<Long, Long>> getCompletedRanges() {
+        List<Tuple<Long, Long>> completedRanges = null;
+        synchronized (mutex) {
+            assert invariant();
+            for (Range range : ranges) {
+                if (range.isPending()) {
+                    continue;
+                }
+                if (completedRanges == null) {
```
Comment (original-brownbear) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:61]: Random NIT/Question: why use a `LinkedList` for this, seems if we're actually worried about this running so often that it'll be tricky on the GC `ArrayList` would perform more predictably and is what we generally use for this type of spot?

**Review by original-brownbear — APPROVED** [review]
LGTM

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/SparseFileTracker.java:60]: Agreed, that's what we usually use, I changed that.

**Comment by tlrx** [comment]
Thanks Armin

**Pull request merged**